akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala

```diff
@@ -0,0 +1,291 @@
+    override def print(t: HttpRequest): Unit = printy(s"\tpushing $t to requestOut")
+  }
+  implicit val ResponseOutPrinterResp = new OpOutPrinter[HttpResponse, Outlet[HttpResponse]] {
+    override def print(t: HttpResponse): Unit = printy(s"\tpushing $t to responseOut")
+  }
+  implicit val RequestInPrinterReq = new OpInPrinter[Inlet[HttpRequest]] {
+    override def print(): Unit = {
+      printy("\tpulling from requestIn")
+    }
+  }
+  implicit val ResponseInPrinterResp = new OpInPrinter[Inlet[HttpResponse]] {
+    override def print(): Unit = {
+      printy("\tpulling from responseIn")
+    }
+  }
+
+  def apply(settings: ClientAutoRedirectSettings) = BidiFlow.fromGraph(new RedirectSupportStage(settings, defaultRedirectMapper))
+
+  def apply(settings: ClientAutoRedirectSettings, redirectMapper: RedirectMapper) = BidiFlow.fromGraph(new RedirectSupportStage(settings, redirectMapper))
+}
+
+/**
+  * A BidiShape that sits on top of OutgoingConnectionBlueprint and routes the redirect responses back to the
+  * downstream outgoing connection if [[redirectMapper]] allows, or upstream otherwise
+  * @param settings Redirect configuration
+  * @param redirectMapper returns Some(HttpRequest) with the redirect HttpRequest if automatic redirect must be
+  *                       performed, or None if received response must go directly upstream
+  */
+class RedirectSupportStage(settings: ClientAutoRedirectSettings, redirectMapper: RedirectMapper) extends GraphStage[BidiShape[HttpRequest, HttpRequest, HttpResponse, HttpResponse]] {
+  /**
+    * in1 is where the original messages come from
+    */
+  val requestIn: Inlet[HttpRequest] = Inlet("Input1")
+
+  /**
+    * in2 is the responses to original messages. If it's a redirect, its not forwarded upstream, but goes downstream
+    */
+  val responseIn: Inlet[HttpResponse] = Inlet("Input2")
+
+  /**
+    * out1 is where input messages go to
+    */
+  val requestOut: Outlet[HttpRequest] = Outlet("Output1")
+
+  /**
+    * out2 is where resulting messages is going to, potentially after redirect
+    */
+  val responseOut: Outlet[HttpResponse] = Outlet("Output2")
+
+  /**
+    * The shape of a graph is all that is externally visible: its inlets and outlets.
+    */
+  override def shape: BidiShape[HttpRequest, HttpRequest, HttpResponse, HttpResponse] =
+    BidiShape(requestIn, requestOut, responseIn, responseOut)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new GraphStageLogic(shape) {
+
+      import RedirectSupportStage.{printy, OpInPrinter, OpOutPrinter, InfiniteRedirectLoopException}
+
+      def doPush[T](outlet: Outlet[T], t: T)(implicit ev: OpOutPrinter[T, Outlet[T]]) = {
+        push(outlet, t)
+        ev.print(t)
+      }
+
+      def doPull[T](inlet: Inlet[T])(implicit ev: OpInPrinter[Inlet[T]]) = {
+        if (!isClosed(inlet)) {
+          pull(inlet)
+        }
+        ev.print()
+      }
+
+      var inFlight = Vector.empty[HttpRequest]
```

Comment:
This is a good question. I'm still not sure if this collection `inFlight` is needed at all with current approach. I've already raised this question [before](https://github.com/akka/akka/issues/15990#issuecomment-198074583). I think, current implementation of the push-pull logic prevents more than 1 request in flight, i.e. I don't pull from `requestIn` before I send the reply to the previous request to the `responseOut`. It is also consistent with the in flight numbers I see during tests.

So with this current approach (always one request in flight), the `inFlight` collection is not needed. I've implemented it like this because I was a bit confused on this topic, but now it starts to become clearer. I think I should either just have 1 item in flight, and keep it in an `Option` instead of `Vector`, or change push-pull logic and truly support multiple, if limited, number of requests in flight. Which approach is preferable?